src/flash.rs

```diff
@@ -2,6 +2,294 @@
+    pub fn page_erase(&mut self, start_offset: u32) -> Result<()> {
+        self.valid_address(start_offset)?;
+
+        // Unlock Flash
+        self.unlock()?;
+
+        // Set Page Erase
+        self.flash.cr.cr().modify(|_, w| w.per().set_bit());
+
+        // Write address bits
+        // NOTE(unsafe) This sets the page address in the Address Register.
+        // The side-effect of this write is that the page will be erased when we
+        // set the STRT bit in the CR below. The address is validated by the
+        // call to self.valid_address() above.
+        unsafe {
+            self.flash
+                .ar
+                .ar()
+                .write(|w| w.far().bits(FLASH_START + start_offset));
+        }
+
+        // Start Operation
+        self.flash.cr.cr().modify(|_, w| w.strt().set_bit());
+
+        // Wait for operation to finish
+        while self.flash.sr.sr().read().bsy().bit_is_set() {}
+
+        // Check for errors
+        let sr = self.flash.sr.sr().read();
+
+        // Remove Page Erase Operation bit
+        self.flash.cr.cr().modify(|_, w| w.per().clear_bit());
+
+        // Re-lock flash
+        self.lock()?;
+
+        if sr.wrprterr().bit_is_set() {
+            self.flash.sr.sr().modify(|_, w| w.wrprterr().clear_bit());
+            Err(Error::EraseError)
+        } else {
+            if self.verify {
+                // By subtracting 1 from the sector size and masking with
+                // start_offset, we make 'start' point to the beginning of the
+                // page, and similarly 'end' at the end of the page. We do this
+                // because the entire page should have been erased, regardless
+                // of where in the page the given 'start_offset' was.
+                let start = start_offset & !(self.sector_sz.kbytes() as u32 - 1);
+                let end = start + self.sector_sz.kbytes() as u32 - 1;
+                for idx in start..end {
+                    let write_address = (FLASH_START + idx as u32) as *const u16;
+                    let verify: u16 = unsafe { core::ptr::read_volatile(write_address) };
+                    if verify != 0xFFFF {
+                        return Err(Error::VerifyError);
+                    }
+                }
+            }
+
+            Ok(())
+        }
+    }
+
+    /// Erase the Flash Sectors from FLASH_START + start_offset to length
+    pub fn erase(&mut self, start_offset: u32, length: usize) -> Result<()> {
+        self.valid_length(start_offset, length)?;
+
+        // Erase every sector touched by start_offset + length
+        for offset in
+            (start_offset..start_offset + length as u32).step_by(self.sector_sz.kbytes() as usize)
+        {
+            self.page_erase(offset)?;
+        }
+
+        // Report Success
+        Ok(())
+    }
+
+    /// Retrieve an slice of data from FLASH_START +offset
+    pub fn read(&self, offset: u32, length: usize) -> Result<&[u8]> {
```

Comment:
So, one downside of this approach is that you have to actually copy flash contents into a user buffer, which isn't actually necessary. An alternative API is as follows:

```
struct MemoryGrant {
    writer: FlashWriter,
    data: &'static [u8],
}

struct FlashWriter {};

impl MemoryGrant {
    pub fn return(self) -> FlashWriter;
}

impl FlashWriter {
    pub fn read(self, address, length) -> MemoryGrant;
}
```

This lets you provide a direct reference to the underlying flash and avoids the copy overhead. The call consumes `self` and stores it in the memory grant. You can then impl `Deref` for the `MemoryGrant` to allow users to directly access the underlying memory I believe.

Reply:
Why do you think we're copying the flash contents ? We're creating a slice directly from flash memory, I think...

Reply:
If that's the case, can't this lead to unsafe behavior? If someone reads from a flash address, but then writes into the data contained there, wouldn't that result in UB? A potentially immutable object contents would be mutated

Reply:
As @thalesfragoso mentions [here](https://github.com/stm32-rs/stm32f1xx-hal/pull/257?fbclid=IwAR33994-VoyGLpFfgZuPe38brCxop-6UXKllIu3RQsnxIvWYo5ArB5u0_eQ#issuecomment-681068488), we can't use the FlashWriter while that slice exists.

Reply:
> we can't use the FlashWriter while that slice exists (the borrow checker will prevent it).

Right now it won't because the slice is marked as `'static`, we need to change that.

Reply:
Although the unsafe block marked the lifetime as static, but my testing indicated the borrow check was still blocking mutable borrows of the `FlashWriter` based on the slice returned by `read()`. I'm curious why that is? In any case, I've modified to explicitly tie the slice to the lifetime of the immutable borrow in `read()`.

Reply:
>  I'm curious why that is?

Hmm, I think I know what was going on, by the lifetime elisions rules, the returned slice would have the same lifetime as the `self` argument, so although we were creating a `'static` slice, its lifetime would shrink to the lifetime of the return type.